### PR TITLE
MRG: scoring in generalization light

### PR DESCRIPTION
**Pull request by kingjr**
Fix scoring functionality in _GeneralizationLight missed in https://github.com/mne-tools/mne-python/pull/3502

### CODE REVIEW

**Comment by kingjr** [comment]
@Eric89GXL I think these are unrelated Travis errors, could you check?

If so, ready for review @kaichogami

**Comment by Eric89GXL** [comment]
Yes they appear unrelated

**Comment by kingjr** [comment]
anyone to review? @jasmainak @dengemann ?

File: mne/decoding/base.py [hunk 1]
```diff
@@ -723,3 +723,25 @@ def _check_estimator(estimator, get_params=True):
...
+    # generated by estimator.score(). Else, we must first get the
+    # predictions based on the scorer.
+    if not isinstance(scoring, str):
+        return make_scorer(scoring) if scoring is not None else scoring
+
+    elif scoring is not None:
+        return get_scorer(scoring)
+    else:
+        raise ValueError(
+            'scoring must be str or scorer got %s instead' % type(scoring))
```
Comment (jasmainak) [inline mne/decoding/base.py]: I'd start a new sentence at "Got %s instead" ...

File: mne/decoding/base.py [hunk 2]
```diff
@@ -723,3 +723,25 @@ def _check_estimator(estimator, get_params=True):
         raise ValueError('estimator must be a scikit-learn transformer or an '
                          'estimator with the get_params method that allows '
                          'cloning.')
+
+
+def _make_scorer(scoring):
+    """Make scorer.
+    Parameters
+    ----------
+    scoring : str | callable
```
Comment (jasmainak) [inline mne/decoding/base.py]: also `None` and instance of `scorer`?

File: mne/decoding/base.py [hunk 3]
```diff
@@ -723,3 +723,25 @@ def _check_estimator(estimator, get_params=True):
         raise ValueError('estimator must be a scikit-learn transformer or an '
                          'estimator with the get_params method that allows '
                          'cloning.')
+
+
+def _make_scorer(scoring):
+    """Make scorer.
```
Comment (jasmainak) [inline mne/decoding/base.py]: Private function but still maybe nicer to add a line after this?

**Comment by jasmainak** [comment]
@kingjr added a few comments but didn't try the changes. Fix looks clean to me

**Comment by agramfort** [comment]
@kingjr can you rebase to make CIs happy?

**Comment by kingjr** [comment]
WIP:I need to add a test test, the make_scorer looks a bit weird.

**Comment by kingjr** [comment]
@agramfort It took me a while to realize that if we do ```cross_val_score(est, X, y, scoring=roc_auc_score)``` it doesn't automatically change the predict_method from `'predict'` to `'predict_proba'` (hence leading to an accuracy score, since AUC on discrete values == accuracy). 

However, it works correctly for ```cross_val_score(est, X, y, scoring='roc_auc')```. We may want to warn the user of this, although perhaps at the sklearn level?

File: mne/decoding/search_light.py [hunk 4]
```diff
@@ -345,7 +336,7 @@ def _sl_score(estimators, scoring, X, y):
     X : array, shape (n_samples, nd_features, n_estimators)
         The target data. The feature dimension can be multidimensional e.g.
         X.shape = (n_samples, n_features_1, n_features_2, n_estimators)
-    scoring : callable, string or None
+    scoring : callable or None
```
Comment (agramfort) [inline mne/decoding/search_light.py]: string not accepted here anymore?

File: mne/decoding/tests/test_search_light.py [hunk 5]
```diff
@@ -57,28 +57,34 @@ def test_SearchLight():
...
+    for err, scoring in [(ValueError, 'foo'), (TypeError, 999)]:
+        sl = _SearchLight(LogisticRegression(), scoring=scoring)
+        sl.fit(X, y)
+        assert_raises(err, sl.score, X, y)
+
+    for method, scoring in [
+            ('predict_proba', 'roc_auc'), ('predict', roc_auc_score)]:
+        sl1 = _SearchLight(LogisticRegression(), scoring=scoring)
+        sl1.fit(X, y)
+        X = np.random.randn(*X.shape)  # randomize X to avoid AUCs in [0, 1]
```
Comment (agramfort) [inline mne/decoding/tests/test_search_light.py]: use a random state

**Comment by agramfort** [comment]
@kingjr just make the docstring clearer. I cannot easily catch such user mistake.

**Comment by agramfort** [comment]
LGTM

now this is code is private no need to update what's new ...

@kingjr at some point let's make it public so you're not the only one to use this? :)

**Comment by kingjr** [comment]
@agramfort it's going to be my personal coding spring

**Comment by Eric89GXL** [comment]
Isn't a "personal coding sprint" otherwise just known as a workday? :)

**Comment by kingjr** [comment]
That's right. On March 26th, I'm planning to work.

**Pull request merged**